src/hardware/bk-1856d/protocol.c

```diff
@@ -0,0 +1,375 @@
+{
+	struct dev_context *devc;
+	struct sr_serial_dev_inst *serial;
+	char *cmd;
+
+	if (!sdi)
+		return;
+
+	devc = sdi->priv;
+	serial = sdi->conn;
+	if (!devc || !serial)
+		return;
+
+	if (devc->sel_input == InputA)
+	{
+		sr_spew("selecting input A");
+		cmd = FUNCTION_A;
+	}
+	else
+	{
+		sr_spew("selecting input C");
+		cmd = FUNCTION_C;
+	}
+
+	if (serial_write_blocking(serial, cmd, LENGHT_OF_CMD,
+			serial_timeout(serial, LENGHT_OF_CMD)) < 1) {
+		sr_err("unable to send function %c command",
+				devc->sel_input == InputA ? 'A' : 'C' );
+	}
+
+	devc->curr_sel_input = devc->sel_input;
+}
+
+static void bk_1856d_chk_select_input(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+
+	if (!sdi)
+		return;
+
+	devc = sdi->priv;
+	if (!devc)
+		return;
+
+	if (devc->sel_input != devc->curr_sel_input)
+		bk_1856d_send_input_sel(sdi);
+}
+
+static void bk_1856d_send_gate_time(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+	struct sr_serial_dev_inst *serial;
+	char *cmd;
+	gulong microseconds;
+
+	if (!sdi)
+		return;
+
+	devc = sdi->priv;
+	serial = sdi->conn;
+	if (!devc || !serial)
+		return;
+
+	if (devc->gate_time < 0) devc->gate_time = 0;
+	if (devc->gate_time > 3) devc->gate_time = 3;
+	switch(devc->gate_time) {
+		default:
+		case 0:
+			cmd = GATE_TIME_0;
+			sr_info("sending gate time 0 (10ms)");
+			microseconds = 40000;
+			break;
+		case 1:
+			cmd = GATE_TIME_1;
+			sr_info("sending gate time 1 (100ms)");
+			microseconds = 80000;
+			break;
+		case 2:
+			cmd = GATE_TIME_2;
+			sr_info("sending gate time 2 (1s)");
+			microseconds = 80000;
+			break;
+		case 3:
+			cmd = GATE_TIME_3;
+			sr_info("sending gate time 3 (10s)");
+			microseconds = 800000;
+			break;
+	}
+
+	if (serial_write_blocking(serial, cmd, LENGHT_OF_CMD,
+			serial_timeout(serial, LENGHT_OF_CMD)) < 1) {
+		sr_err("unable to send gate time command");
+	}
+	g_usleep(microseconds);
+}
+
+static void bk_1856d_request_data(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+	struct sr_serial_dev_inst *serial;
+
+	if (!sdi)
+		return;
+
+	devc = sdi->priv;
+	serial = sdi->conn;
+	if (!devc || !serial)
+		return;
+
+	sr_spew("requesting data");
+
+	if (serial_write_blocking(serial, DATA_REQ, LENGHT_OF_CMD,
+			serial_timeout(serial, LENGHT_OF_CMD)) < 1) {
+		sr_err("unable to send request data command");
+	}
+}
+
+SR_PRIV void bk_1856d_init(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+	struct sr_serial_dev_inst *serial;
+
+	if (!sdi)
+		return;
+
+	devc = sdi->priv;
+	serial = sdi->conn;
+	if (!devc || !serial)
+		return;
+
+	devc->buffer_level = 0;
+	sr_sw_limits_acquisition_start(&(devc->sw_limits));
+	serial_flush(serial);
+
+	g_mutex_lock(&devc->rw_mutex);
+	bk_1856d_send_input_sel(sdi);
+
+	bk_1856d_send_gate_time(sdi);
+	bk_1856d_request_data(sdi);
+	g_mutex_unlock(&devc->rw_mutex);
+}
+
+static int bk_1856d_check_for_zero_message(struct dev_context *devc)
```

Comment:
it would be helpful to the reviewer (present and future) to give a few examples, in a comment, of typical strings and expected return values.

Reply:
I added a link to the vendors manual.